packages/server/src/accounts-server.ts

```diff
@@ -528,6 +577,40 @@ Please change it with a strong random token.`);
       ? this.options.tokenCreator.createToken(user)
       : generateRandomToken();
   }
+
+  private async getUserFromMfaToken({
+    mfaToken,
+    loginToken,
+    skipValidation = false,
+  }: {
+    mfaToken: string;
+    loginToken?: string;
+    skipValidation: boolean;
+  }): Promise<User | null> {
+    const loginAttempt = await this.db.getMfaLoginAttempt(mfaToken);
+
+    if (!loginAttempt) {
+      return null;
+    }
+
+    if (!skipValidation && loginAttempt.loginToken !== loginToken) {
+      return null;
+    }
+
+    return this.db.findUserById(loginAttempt.userId);
+  }
+
+  private async createMfaLoginProcess(user: User): Promise<MFALoginResult> {
+    const loginToken = generateRandomToken();
```

Comment:
@ozsay I don't exactly understand why do we need this `loginToken`, can you explain a bit why is it really needed please? :)

Reply:
the user can only log in if he has the `loginToken` which is provided to him upon successfully completing the challenge.

Did I over-thought this? :)

Reply:
Since the mfaToken is already a random token I feel like having this `loginToken` does not add more security no?

Reply:
following `auth0`, the last step of the mfa is to call `oauth/token` (in our case `loginWithService`) with a token. I can't use the `mfaToken` here because it's provided to the client before it does the challenge. that's why I have the extra token.

if we want, we can skip this step and have the `performMfaChallenge` create the session and return the access and refresh tokens.